format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/PathTrackingDelimitedInputFormat.java

```diff
@@ -120,4 +137,77 @@ public void close() throws IOException {
       }
     };
   }
+
+  /**
+   * Split the delimited string based on the delimiter. The delimiter should not contain any quotes. The method will
+   * behave like this:
+   * 1. if there is no quote, it will behave same as {@link String#split(String)}
+   * 2. if there are quotes in the string, the method will find pairs of quotes, content within each pair of quotes
+   * will not get splitted even if there is delimiter in that.
+   * For example, if string is "a.\"b.c\".\"d.e.f\"" and delimiter is '.', it will get split into [a, b.c, d.e.f].
+   * if string is "val1\".\"val2\"", then it will not get splitted since the '.' is within pair of quotes.
+   *
+   * @param delimitedString the string to split
+   * @param delimiter the separtor
+   * @return a list of splits of the original string
+   */
+  @VisibleForTesting
+  static List<String> splitQuotesString(String delimitedString, String delimiter) {
+    // use a tree map so we can easily look up if a delimiter is within a pair of quotes
+    // the map has key of index and value to the index of paired quote.
+    TreeMap<Integer, Integer> quotesPos = new TreeMap<>();
+    int firstQuotePos = -1;
+    for (int i = 0; i < delimitedString.length(); i++) {
+      if (delimitedString.charAt(i) == '\"') {
+        if (firstQuotePos != -1) {
+          quotesPos.put(i, firstQuotePos);
+          quotesPos.put(firstQuotePos, i);
+          firstQuotePos = -1;
+        } else {
+          firstQuotePos = i;
+        }
+      }
+    }
+
+    List<String> result = new ArrayList<>();
+    StringBuilder split = new StringBuilder();
+    for (int i = 0; i < delimitedString.length(); i++) {
+      // if the length is not enough for the delimiter, just add it to split
+      if (i + delimiter.length() > delimitedString.length()) {
+        split.append(delimitedString.charAt(i));
+      } else {
+        // find a delimiter
+        if (delimitedString.substring(i, i + delimiter.length()).equals(delimiter)) {
+          // find the prev quote and next quote index
+          Integer prevQuotePos = quotesPos.floorKey(i);
+          Integer nextQuotePos = quotesPos.ceilingKey(i);
+          // if the they exist and they form a pair, we will not split, just add the current character to split
+          if (prevQuotePos != null && nextQuotePos != null && quotesPos.get(prevQuotePos).equals(nextQuotePos)) {
```

Comment:
I'm a bit confused here. Say the line is:
```
pos: 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 |
     " | a | , | b | " | , | " | c | , | d | "  |
```

that means quotesPos is:

0 -> 4
6 -> 10

So doesn't this mean it mean `quotesPos.get(prevQuotePos)` will never be the same as `nextQuotePos`?

Reply:
ok just realized the map has both left -> right as well as right -> left, so it would be:

```
0 -> 4
4 -> 0
6 -> 10
10 -> 6
```

Can you add more comments with an example?